crates/kitsune_p2p/proxy/src/tx2.rs

```diff
@@ -0,0 +1,598 @@
+        }
+        .boxed()
+    }
+}
+
+async fn incoming_evt_logic(
+    mut sub_ep: Ep,
+    hnd: Arc<ProxyEpHnd>,
+    local_digest: CertDigest,
+    logic_hnd: LogicChanHandle<EpEvent>,
+) {
+    while let Some(evt) = sub_ep.next().await {
+        use EpEvent::*;
+        if match evt {
+            IncomingConnection(_) => Ok(()),
+            IncomingData(EpIncomingData {
+                con: sub_con,
+                url: base_url,
+                msg_id,
+                mut data,
+            }) => {
+                if data.is_empty() {
+                    // TODO - FIXME - kill connection
+                    panic!("corrupt incoming data");
+                }
+                match data[0] {
+                    PROXY_HELLO_DIGEST => {
+                        let remote_digest = CertDigest(Arc::new(data[1..].to_vec()));
+                        ingest_incoming_con(
+                            logic_hnd.clone(),
+                            hnd.inner.clone(),
+                            local_digest.clone(),
+                            remote_digest,
+                            base_url,
+                            sub_con,
+                        )
+                        .await;
+                    }
+                    PROXY_FWD_MSG => {
+                        let src_digest = CertDigest(Arc::new(data[33..65].to_vec()));
+                        let dest_digest = CertDigest(Arc::new(data[1..33].to_vec()));
+                        if dest_digest == hnd.cert_digest {
+                            // this data is destined for US!
+                            data.cheap_move_start(65);
+                            let url = match promote_addr(&base_url, &src_digest) {
+                                Ok(url) => url,
+                                // TODO - FIXME
+                                e => panic!("{:?}", e),
+                            };
+                            let con = ProxyConHnd::new(sub_con, dest_digest, src_digest);
+                            let evt = EpEvent::IncomingData(EpIncomingData {
+                                con,
+                                url,
+                                msg_id,
+                                data,
+                            });
+                            let _ = logic_hnd.emit(evt).await;
+                        } else {
+                            let dest = hnd.inner.share_mut(|i, _| {
+                                Ok(i.in_digest_to_sub_con.get(&dest_digest).cloned())
+                            });
+                            match dest {
+                                Ok(Some(d_sub_con)) => {
+                                    let t = KitsuneTimeout::from_millis(1000 * 30);
+                                    // TODO - correct to await here / backpressure?
+                                    // TODO - respond with errors?
+                                    let _ = d_sub_con.write(msg_id, data, t).await;
```

Comment:
Does this make use single threaded for all proxy messages?

Reply:
Yes. Interestingly enough, performance seems to degrade slightly when I run this logic with max cpu_count semaphore tasks... 

But there is the possibility of backed up writes slowing us down, and we've got 3 channels to work with, so I switched this to use `for_each_concurrent(3, ...` which doesn't seem to have any effect on performance for our benchmark, but could theoretically speed up that back-pressured writer case?